pkg/apis/ops/v1alpha1/types.go

```diff
@@ -103,9 +103,15 @@ type Traceflow struct {
 
 // TraceflowSpec describes the spec of the traceflow.
 type TraceflowSpec struct {
-	Source      Source      `json:"source,omitempty"`
-	Destination Destination `json:"destination,omitempty"`
-	Packet      Packet      `json:"packet,omitempty"`
+	Source      Source       `json:"source,omitempty"`
+	Destination *Destination `json:"destination,omitempty"`
+	Packet      *Packet      `json:"packet,omitempty"`
```

Comment:
I do not believe this is considered a backwards-compatible change unfortunately, so the API version needs to change. A Go client for example would need to be updated.

Reply:
I do feel these changes are compatible. For example, I can test with old antctl + old controller + new agent.

Reply:
The serialized data is the same indeed. Let me see if I find something in the K8s API documentation about changing from non-pointer to pointer type.

Reply:
Sure. Let me know. If it is not compatible, I might choose to revert the changes.

Reply:
All I found with a quick search was [this](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility):

> An API change is considered compatible if it [...] does not change existing semantics, including [...] which fields are required and which are not

And later on in the same document, when discussing upgrading alpha APIs:

> developers should still increment the API version when object schema or semantics change in an incompatible way

I think typically making a field optional would be considered incompatible because a client may list all resources and fail to process them (e.g. segfault) if a field that was previously required is now optional (i.e. became a pointer). Because of the nature of the Traceflow API, its current (known) consumers, and the usage these consumers make of the API, I would agree that it is an unlikely situation. If I am writing a Go client which uses an API like Traceflow, it is surprising that the package remains exactly the same (`pkg/apis/ops/v1alpha1`), yet my code has to change if I want to be able to keep building it with the latest imported code. If we do want to apply the guidelines for API versioning, it seems to me we should still increase the version to v1alpha2. What do you think @tnqn?

Reply:
I feel changing from required to not is better...

I do not know if code level compatibility is also covered by API version. Like to learn.

Reply:
> I feel changing from required to not is better...

For a POST yes, but for a GET?

> I do not know if code level compatibility is also covered by API version. Like to learn.

I feel like they go hand-in-hand here, if you consider API reads in addition to writes

Reply:
Ok. Let us see what @tnqn thinks. Given we are making APIGroup changes and there are only several days left for the release, I would just revert the changes if we need to add a new API version for them.

Reply:
> I do feel these changes are compatible. For example, I can test with old antctl + old controller + new agent.

The scenario we need to support is new controller + old agent? I'm thinking a case that new antctl create a Traceflow that doesn't set Destination and Packet, would agent crash or the optional fields will be filled up with zero value?
As @antoninbas's link pointed out, changing existing fields' semantics in this way is considered incompatible and is generally disallowed. I don't know what it would break except the code consuming it. But if it's just good to make it optional, not must, I feel it's better to defer the change to next API version to be riskless.

Reply:
I think old agent will not crash, as its go client will populate an empty struct.

Ok, as we believe generally changing to optional is an incompatible change, let me drop it in this release.